lib/npg_pipeline/executor/wr.pm

```diff
@@ -0,0 +1,236 @@
+         map { $json->encode($_) } # convert every wr definition to JSON
+         map { @{$_} }             # expand arrays of wr definitions        
+         values %{$self->commands4jobs()}
+                               );
+    $action = 'submitting';
+    $self->_submit();
+  } catch {
+    $self->logcroak(qq[Error $action wr jobs: $_]);
+  };
+
+  return;
+};
+
+##################################################################
+############## Private methods ###################################
+##################################################################
+
+sub _process_function {
+  my ($self, $function) = @_;
+
+  my $g = $self->function_graph4jobs;
+  #####
+  # Need dependency group ids of upstream functions in order to set correctly
+  # dependencies between wr jobs
+  #
+  my @depends_on = ();
+  if (!$g->is_source_vertex($function)) {
+    @depends_on = $self->dependencies($function, $VERTEX_GROUP_DEP_ID_ATTR_NAME);
+    if (!@depends_on) {
+      $self->logcroak(qq["$function" should depend on at least one job]);
+    }
+  }
+
+  my $group_id = $self->_definitions4function($function, @depends_on);
+  if (!$group_id) {
+    $self->logcroak(q[Group dependency id should be returned]);
+  }
+
+  $g->set_vertex_attribute($function, $VERTEX_GROUP_DEP_ID_ATTR_NAME, $group_id);
+
+  return;
+}
+
+sub _definitions4function {
+  my ($self, $function_name, @depends_on) = @_;
+
+  my $definitions = $self->function_definitions()->{$function_name};
+  my $group_id = join q[-], $function_name, $definitions->[0]->identifier(), irand();
+
+  foreach my $d (@{$definitions}) {
+    my $wr_def = $self->_definition4job($function_name, $d);
+    if (@depends_on) {
+      $wr_def->{'deps'} = \@depends_on;
+    }
+    $wr_def->{'dep_grps'} = [$group_id];
+    $wr_def->{'rep_grp'}  = join q[-], $d->identifier(), $function_name;
+    push @{$self->commands4jobs()->{'function_name'}}, $wr_def;
+  }
+
+  return $group_id;
+}
+
+sub _definition4job {
+  my ($self, $function_name, $d) = @_;
+
+  my $def = {};
+
+  $def->{'memory'} = $d->has_memory() ? $d->memory() : $DEFAULT_MEMORY;
+  $def->{'memory'} .= q[M];
+
+  if ($d->has_num_cpus()) {
+    use warnings FATAL => qw(numeric);
+    $def->{'cpus'} = int $d->num_cpus()->[0];
+  }
+
+  my $log_file = sub {
+    my $log_name = join q[-], $function_name, $d->created_on(),
+      $d->has_composition() ? $d->composition()->freeze2rpt () : $d->identifier();
+    $log_name   .= q[.out];
+    return join q[/], $d->log_file_dir(), $log_name;
+  };
+
+  $def->{'cmd'} = join q[ ], q[(], $d->command(), q[)], q[2>&1], q[|], q[tee], $log_file->();
+
+  return $def;
+}
+
+sub _wr_add_command {
+  my $self = shift;
+
+  # If needed, in future, these options can be passed from the command line
+  # or read fron a conf. file.
+  my @common_options = (
+          '--cwd'        => '/tmp',
+          '--disk'       => 0,
+          '--override'   => 2,
+          '--priority'   => 50,
+          '--retries'    => 1,
```

Comment:
Do you want an automatic retry on failure, or did you mean 0, for no automatic retries?